src/util/lazy.h

```diff
@@ -0,0 +1,46 @@
+/*******************************************************************\
+
+Module: Util
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_LAZY_H
+#define CPROVER_UTIL_LAZY_H
+
+#include <functional>
+#include <util/optional.h>
+
+template <typename valuet, typename infot>
+class lazyt
+{
+public:
+  /// Delay the computation of \p fun to the next time the \c force method
+  /// is called.
+  static lazyt from_fun(std::function<valuet(infot)> fun)
+  {
+    return lazyt{std::move(fun)};
+  }
```

Comment:
Why not a public constructor?

Reply:
I think having a static function with a name is clearer in that case, when I read `lazyt::from_fun(f)` I know we use a function `f` to make a `lazyt` object, but `lazyt(f)` is more obscure.